### ISSUE TITLE
Accessibility: Rust Analyzer in VSCode doesn't allow reading variable types using screenreader

### ISSUE THREAD

**Issue by denizsincar29** [rust-lang/rust-analyzer#18158]
Hello. I don't know if it's the issue of rust analyzer or VSCode itself, but I just now realized that variable types are written near every variable, (that one sighted guy told me). I'm blind and I use a screenreader called NVDA. VSCode is the most accessible IDE ever, but not all of the extensions share the same accessibility. My screenreader doesn't announce variable types when cursoring to a variable.

RustAnalyzer extension is set to autoupdate, I use the latest for 21 september 2024.
The rustc version is 1.81.0,.
I use the last for this date version of VSCode.

**Comment by lnicola** [comment]
These are a VS Code feature called inlay hints. We only send the hint text, and Code renders it.

According to https://github.com/microsoft/vscode/issues/142532, there's a command (F1, Read Line with Inlay Hint) that reads them, which makes me think they aren't read by default. I can't easily test, as I'm not running Windows and I have no idea how to set up and use a screen reader, but please give that a try first.

**Comment by denizsincar29** [comment]
I tested it and couldn't get how they work. My screenreader stucks there.

**Comment by lnicola** [comment]
@jrieken sorry for the ping, but do you happen to know if there's any way we can mess up this feature?

**Comment by denizsincar29** [comment]
Checked the issue and this thing doesn't work as needed or I couldn't get it. Maybe you could add something like an accessible view that shows types of variables on the current line as a temporary solution?

**Comment by denizsincar29** [comment]
The only workaround I see is to delete the last letter of a variable name and rewrite this letter, causing the suggestion to appear. It will suggest this variable with it's type. Than you will just close it by pressing escape. In functions, when you type parameters, it screenreads the docs of each parameter. In structs the same thing.

**Comment by lnicola** [comment]
Another workaround might be to do F1 / Show or Focus hover, which pops up a panel with the type, memory layout (which you can disable using `rust-analyzer.hover.memoryLayout.enable`) and some actions.

**Comment by tgrushka** [comment]
Hi guys, on macOS, VoiceOver reads the inline hints for me when I choose "Read Line With Inline Hints" command. I agree this is annoying. I am legally blind and use a custom screen reader I wrote that is based on copy-paste, and VS Code has no equivalent of "Copy Line With Inline Hints," which it should, but VoiceOver does read it for me, even with `"editor.inlayHints.enabled": "offUnlessPressed"` in settings.

I will try on my Windows and Linux machines as soon as I get the chance.

For rust-analyzer devs on macOS:
1. Ensure "Screen Reader Optimized" is visible in status bar. If not, run command: "Toggle Screen Reader Accessibility Mode."
2. Focus on the line with inline hints in VS Code to test.
3. Press Command-F5 to turn on VoiceOver. If you're not used to using a screen reader, navigation is very different, hence I suggest focusing on the relevant code line first.
4. Press up arrow, listen, press down arrow, listen again. Note the lines are read without inline hints.
5. Choose "Read Line with Inline Hints" command. It should read out with inline hints.
6. Press Command-F5 to turn off VoiceOver.

I will try soon on Windows NVDA on my other laptop.

**Comment by jrieken** [comment]
> EDIT: VS Code really should have an option to toggle this on/off "permanently."

@tgrushka It has - the `editor.inlayHints.enabled` setting also supports `on` and `off` values


I also suggest we move this issue over into the vscode-repo because the rust extension cannot do much about this.

**Comment by denizsincar29** [comment]
Wow, i found a solution. To see a very good understanding on basically everything under cursor, you press ctrl+k and then i, and than press alt+f2 to accessible view.
It can even tell me, what does the word let, mut, struct, fn... mean. It says the type and i guess, stack size of a variable, documentation of some functions and some other great stuff!
